### PR TITLE
Push down dereference expression

### PR DESCRIPTION
**Pull request by zhenxiao**
Co-authored-by: qqibrow qqibrow@gmail.com

To fix https://github.com/prestodb/presto/issues/14517, we need to pushdown dereference expressions as a first step.

continue https://github.com/prestodb/presto/pull/5547 https://github.com/prestodb/presto/pull/13180

```
== RELEASE NOTES ==

General Changes
* push down dereference expression

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/PushDownDereferences.java [hunk 1]
```diff
@@ -0,0 +1,560 @@
...
+        @Override
+        protected FilterNode rewrite(Context context, FilterNode node, BiMap<DereferenceExpression, VariableReferenceExpression> expressions)
+        {
+            PlanNode source = node.getSource();
+
+            ImmutableMap<VariableReferenceExpression, RowExpression>
+                    dereferencesMap =
+                    expressions.inverse().entrySet().stream().collect(toImmutableMap(Map.Entry::getKey,
+                            entry -> castToRowExpression(entry.getValue())));
+            Assignments assignments = Assignments.builder().putAll(identityAssignmentsAsSymbolReferences(source.getOutputVariables())).putAll(dereferencesMap).build();
```
Comment (vkorukanti) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/PushDownDereferences.java:182]: For this case:
Filter(predicate(a.x) = 5, output = a.y)
   Project(a)
-->
Filter(predicate(a_x) = 5, output = a_y)
   Project(a, deref(a.x) as a_x, deref(a.y) as a_y)

Is the new project correct, given that we are adding the existing project symbols and the deref expressions as new projects?

File: presto-main/src/test/java/com/facebook/presto/sql/planner/TestDereferencePushDown.java [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+import static com.facebook.presto.sql.planner.assertions.PlanMatchPattern.unnest;
+import static com.facebook.presto.sql.planner.assertions.PlanMatchPattern.values;
+import static com.facebook.presto.sql.planner.plan.ExchangeNode.Scope.LOCAL;
+import static com.facebook.presto.sql.planner.plan.ExchangeNode.Type.GATHER;
+import static com.facebook.presto.sql.planner.plan.ExchangeNode.Type.REPARTITION;
+import static com.facebook.presto.sql.planner.plan.JoinNode.Type.INNER;
+import static com.facebook.presto.sql.tree.SortItem.NullOrdering.LAST;
+import static com.facebook.presto.sql.tree.SortItem.Ordering.ASCENDING;
+
+public class TestDereferencePushDown
```
Comment (vkorukanti) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/TestDereferencePushDown.java:43]: Can we also run these tests with `experimental.pushdown-subfields-enabled=true`? This way we make sure when both of these are enabled, we don't get into any issues.

File: presto-main/src/test/java/com/facebook/presto/sql/planner/TestDereferencePushDown.java [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+                                                values("msg"))))));
+
+        assertPlan("WITH t(msg) AS ( SELECT * FROM (VALUES ROW(CAST(ROW(1, 2.0) AS ROW(x BIGINT, y DOUBLE))))) " +
+                        "SELECT a.msg.y FROM t a JOIN t b ON a.msg.y = b.msg.y WHERE a.msg.x > bigint '5'",
+                output(ImmutableList.of("a_y"),
+                        join(INNER, ImmutableList.of(equiJoinClause("a_y", "b_y")),
+                                anyTree(
+                                        project(ImmutableMap.of("a_y", expression("msg.y")),
+                                                filter("msg.x > bigint '5'",
+                                                        values("msg")))
```
Comment (vkorukanti) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/TestDereferencePushDown.java:67]: Shouldn't we have a `Project(msg.x as a_x, msg.y as a_y) ` in between the `filter` and `values` nodes?

File: presto-main/src/test/java/com/facebook/presto/sql/planner/TestDereferencePushDown.java (hunk 2 above)
Comment (zhenxiao) [reply presto-main/src/test/java/com/facebook/presto/sql/planner/TestDereferencePushDown.java:44]: yep, will fix

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/PushDownDereferences.java [hunk 4]
```diff
@@ -0,0 +1,560 @@
...
+
+    /**
+     * Extract dereferences and push them down to new ProjectNode below
+     * Transforms:
+     * <pre>
+     *  TargetNode(expression(a.x))
+     *  </pre>
+     * to:
+     * <pre>
+     *   ProjectNode(original symbols)
```
[the remark this answers is not in the document]
Comment (zhenxiao) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/PushDownDereferences.java]: a projectNode on top is to keep all upper symbols unchanged

File: presto-main/src/test/java/com/facebook/presto/sql/planner/TestDereferencePushDown.java (hunk 3 above)
Comment (zhenxiao) [reply presto-main/src/test/java/com/facebook/presto/sql/planner/TestDereferencePushDown.java:73]: image

only a_y appears in top ProjectNode, so a_y is pushed down. a_x only appears in FilterNode, it is not pushed down.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushDownDereferences.java [hunk 5]
```diff
@@ -0,0 +1,560 @@
...
+        public Pattern<N> getPattern()
+        {
+            return Pattern.typeOf(planNodeClass);
+        }
+
+        @Override
+        public Result apply(N node, Captures captures, Context context)
+        {
+            Map<DereferenceExpression, VariableReferenceExpression> expressions =
+                    getDereferenceSymbolMap(ExpressionExtractor.extractExpressionsNonRecursive(node), context, metadata, sqlParser);
```
Comment (mbasmanova) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushDownDereferences.java:153]: static import ExpressionExtractor.extractExpressionsNonRecursive

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushDownDereferences.java [hunk 6]
```diff
@@ -0,0 +1,560 @@
...
+        return builder.build();
+    }
+
+    private static Map<DereferenceExpression, VariableReferenceExpression> getDereferenceSymbolMap(Collection<RowExpression> expressions, Context context, Metadata metadata, SqlParser sqlParser)
+    {
+        Set<DereferenceExpression> dereferences = expressions.stream()
+                .filter(OriginalExpressionUtils::isExpression)
+                .map(OriginalExpressionUtils::castToExpression)
+                .flatMap(expression -> extractDereferenceExpressions(expression).stream())
+                .filter(PushDownDereferences::validPushDown)
```
Comment (mbasmanova) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushDownDereferences.java:518]: Looks like validPushDown logic can be folded into extractDereferenceExpressions, e.g. extractDereferenceExpressions can be modified to extract only "valid" dereferences.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushDownDereferences.java [hunk 7]
```diff
@@ -0,0 +1,560 @@
...
+                .collect(toImmutableSet());
+
+        // TODO DereferenceExpression Base will cause unnecessary rewritten of same expression.
```
Comment (mbasmanova) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushDownDereferences.java:521]: I'm not sure I understand this TODO and the logic in the following if statement. Looks like [msg.foo, msg.foo.bar] can is not being processed at all. Is this intentional? Shouldn't this case result in pushing down msg.foo?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushDownDereferences.java [hunk 8]
```diff
@@ -0,0 +1,560 @@
...
+        {
+            super(FilterNode.class);
+        }
+
+        @Override
+        protected FilterNode rewrite(Context context, FilterNode node, BiMap<DereferenceExpression, VariableReferenceExpression> expressions)
+        {
+            PlanNode source = node.getSource();
+
+            ImmutableMap<VariableReferenceExpression, RowExpression>
```
Comment (mbasmanova) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushDownDereferences.java:178]: - ImmutableMap -> Map
- formatting

```
            Map<VariableReferenceExpression, RowExpression> dereferencesMap = expressions.inverse().entrySet().stream()
                            .collect(toImmutableMap(Map.Entry::getKey, entry -> castToRowExpression(entry.getValue())));
            Assignments assignments = Assignments.builder()
                    .putAll(identityAssignmentsAsSymbolReferences(source.getOutputVariables()))
                    .putAll(dereferencesMap)
                    .build();
```

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushDownDereferences.java [hunk 9]
```diff
@@ -0,0 +1,560 @@
...
+            return Result.ofPlanNode(newUnnest);
+        }
+    }
+
+    private static PlanNode createProjectBelow(PlanNode planNode, Assignments dereferences, PlanNodeIdAllocator idAllocator)
+    {
+        if (dereferences.isEmpty()) {
+            return planNode;
+        }
+        return new ProjectNode(idAllocator.getNextId(), planNode, Assignments.builder().putAll(identityAssignmentsAsSymbolReferences(planNode.getOutputVariables())).putAll(dereferences).build());
```
Comment (mbasmanova) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushDownDereferences.java:474]: perhaps, extract assignments variable for readability:

```
        Assignments assignments = Assignments.builder()
                .putAll(identityAssignmentsAsSymbolReferences(planNode.getOutputVariables()))
                .putAll(dereferences)
                .build();
        return new ProjectNode(idAllocator.getNextId(), planNode, assignments);
```

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushDownDereferences.java [hunk 10]
```diff
@@ -0,0 +1,560 @@
...
+            UnnestNode newUnnest = new UnnestNode(context.getIdAllocator().getNextId(),
+                    source,
+                    ImmutableList.<VariableReferenceExpression>builder().addAll(unnestNode.getReplicateVariables()).addAll(expressions.values()).build(),
+                    unnestNode.getUnnestVariables(),
+                    unnestNode.getOrdinalityVariable());
+            return Result.ofPlanNode(newUnnest);
+        }
+    }
+
+    private static PlanNode createProjectBelow(PlanNode planNode, Assignments dereferences, PlanNodeIdAllocator idAllocator)
```
Comment (mbasmanova) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushDownDereferences.java:469]: this name is confusing to me; the implementation adds ProjectNode above the input planNode, not below

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushDownDereferences.java [hunk 11]
```diff
@@ -0,0 +1,560 @@
...
+            PlanNode source = node.getSource();
+
+            ImmutableMap<VariableReferenceExpression, RowExpression>
+                    dereferencesMap =
+                    expressions.inverse().entrySet().stream().collect(toImmutableMap(Map.Entry::getKey,
+                            entry -> castToRowExpression(entry.getValue())));
+            Assignments assignments = Assignments.builder().putAll(identityAssignmentsAsSymbolReferences(source.getOutputVariables())).putAll(dereferencesMap).build();
+            ProjectNode projectNode = new ProjectNode(context.getIdAllocator().getNextId(), source, assignments);
+
+            RowExpression filter = castToRowExpression(ExpressionTreeRewriter.rewriteWith(new DereferenceReplacer(expressions), castToExpression(node.getPredicate())));
```
Comment (mbasmanova) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushDownDereferences.java:185]: this logic repeats in 3 places; consider extracting helper method:

```
    private RowExpression replaceDereferences(RowExpression rowExpression, Map<DereferenceExpression, VariableReferenceExpression> dereferences) {
        return castToRowExpression(rewriteWith(new DereferenceReplacer(dereferences), castToExpression(rowExpression)));
    }

            return new FilterNode(
                    context.getIdAllocator().getNextId(),
                    projectNode,
                    replaceDereferences(node.getPredicate(), expressions));
```

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushDownDereferences.java [hunk 12]
```diff
@@ -0,0 +1,560 @@
...
+            PlanNode rightNode = createProjectBelow(joinNode.getRight(), rightSideDereferences.build(), context.getIdAllocator());
+
+            return new JoinNode(
+                    context.getIdAllocator().getNextId(),
+                    joinNode.getType(),
+                    leftNode,
+                    rightNode,
+                    joinNode.getCriteria(),
+                    ImmutableList.<VariableReferenceExpression>builder().addAll(leftNode.getOutputVariables()).addAll(rightNode.getOutputVariables()).build(),
+                    joinNode.getFilter().map(expression -> castToRowExpression(ExpressionTreeRewriter.rewriteWith(new PushDownDereferences.DereferenceReplacer(expressions), castToExpression(expression)))),
```
Comment (mbasmanova) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushDownDereferences.java:227]: use a helper method from above:

```
                    joinNode.getFilter().map(expression -> replaceDereferences(expression, expressions)),
```

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushDownDereferences.java [hunk 13]
```diff
@@ -0,0 +1,560 @@
...
+            PlanNode leftNode = createProjectBelow(joinNode.getLeft(), leftSideDereferences.build(), context.getIdAllocator());
+            PlanNode rightNode = createProjectBelow(joinNode.getRight(), rightSideDereferences.build(), context.getIdAllocator());
+
+            return new JoinNode(
+                    context.getIdAllocator().getNextId(),
+                    joinNode.getType(),
+                    leftNode,
+                    rightNode,
+                    joinNode.getCriteria(),
+                    ImmutableList.<VariableReferenceExpression>builder().addAll(leftNode.getOutputVariables()).addAll(rightNode.getOutputVariables()).build(),
```
Comment (mbasmanova) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushDownDereferences.java:226]: formatting

```
                    ImmutableList.<VariableReferenceExpression>builder()
                            .addAll(leftNode.getOutputVariables())
                            .addAll(rightNode.getOutputVariables())
                            .build(),
```

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushDownDereferences.java [hunk 14]
```diff
@@ -0,0 +1,560 @@
...
+            return project().with(source().matching(targetPattern.capturedAs(targetCapture)));
+        }
+
+        @Override
+        public Result apply(ProjectNode node, Captures captures, Context context)
+        {
+            N child = captures.get(targetCapture);
+            Map<DereferenceExpression, VariableReferenceExpression> allDereferencesInProject = getDereferenceSymbolMap(node.getAssignments().getExpressions(), context, metadata, sqlParser);
+
+            Set<VariableReferenceExpression> childSourceSymbols = child.getSources().stream().map(PlanNode::getOutputVariables).flatMap(Collection::stream).collect(toImmutableSet());
```
Comment (mbasmanova) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushDownDereferences.java:260]: formatting

```
            Set<VariableReferenceExpression> childSourceSymbols = child.getSources().stream()
                    .map(PlanNode::getOutputVariables).flatMap(Collection::stream)
                    .collect(toImmutableSet());
```

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushDownDereferences.java [hunk 15]
```diff
@@ -0,0 +1,560 @@
...
+        {
+            super(Pattern.typeOf(planNodeClass));
+        }
+
+        @Override
+        protected Result pushDownDereferences(Context context, N targetNode, BiMap<DereferenceExpression, VariableReferenceExpression> expressions)
+        {
+            PlanNode source = getOnlyElement(targetNode.getSources());
+
+            ImmutableMap<VariableReferenceExpression, RowExpression>
```
Comment (mbasmanova) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushDownDereferences.java:316]: - ImmutableMap -> Map
- formatting

```
            Map<VariableReferenceExpression, RowExpression> dereferencesMap =
                    expressions.inverse().entrySet().stream()
                            .collect(toImmutableMap(Map.Entry::getKey, entry -> castToRowExpression(entry.getValue())));
            ProjectNode projectNode = new ProjectNode(
                    context.getIdAllocator().getNextId(),
                    source,
                    Assignments.builder()
                            .putAll(identityAssignmentsAsSymbolReferences(source.getOutputVariables()))
                            .putAll(dereferencesMap)
                            .build());
```

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushDownDereferences.java [hunk 16]
```diff
@@ -0,0 +1,560 @@
...
+            PlanNode leftNode = createProjectBelow(joinNode.getLeft(), leftSideDereferences.build(), context.getIdAllocator());
+            PlanNode rightNode = createProjectBelow(joinNode.getRight(), rightSideDereferences.build(), context.getIdAllocator());
+
+            return Result.ofPlanNode(new JoinNode(
+                    context.getIdAllocator().getNextId(),
+                    joinNode.getType(),
+                    leftNode,
+                    rightNode,
+                    joinNode.getCriteria(),
+                    ImmutableList.<VariableReferenceExpression>builder().addAll(leftNode.getOutputVariables()).addAll(rightNode.getOutputVariables()).build(),
```
Comment (mbasmanova) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushDownDereferences.java:380]: formatting

```
                    ImmutableList.<VariableReferenceExpression>builder()
                            .addAll(leftNode.getOutputVariables())
                            .addAll(rightNode.getOutputVariables())
                            .build(),
```

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushDownDereferences.java [hunk 17]
```diff
@@ -0,0 +1,560 @@
...
+    {
+        PushDownDereferenceThroughProject()
+        {
+            super(project());
+        }
+
+        @Override
+        protected Result pushDownDereferences(Context context, ProjectNode projectNode, BiMap<DereferenceExpression, VariableReferenceExpression> expressions)
+        {
+            ImmutableMap<VariableReferenceExpression, RowExpression>
```
Comment (mbasmanova) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushDownDereferences.java:428]: formatting

```
            Map<VariableReferenceExpression, RowExpression> dereferencesMap = expressions.inverse().entrySet().stream()
                    .collect(toImmutableMap(Map.Entry::getKey, entry -> castToRowExpression(entry.getValue())));

            return Result.ofPlanNode(
                    new ProjectNode(context.getIdAllocator().getNextId(),
                            projectNode.getSource(),
                            Assignments.builder()
                                    .putAll(projectNode.getAssignments())
                                    .putAll(dereferencesMap)
                                    .build()));
```

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushDownDereferences.java [hunk 18]
```diff
@@ -0,0 +1,560 @@
...
+    {
+        PushDownDereferenceThroughUnnest()
+        {
+            super(unnest());
+        }
+
+        @Override
+        protected Result pushDownDereferences(Context context, UnnestNode unnestNode, BiMap<DereferenceExpression, VariableReferenceExpression> expressions)
+        {
+            // Create new Project contains all pushdown symbols above original source
```
Comment (mbasmanova) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushDownDereferences.java:451]: - remove comments, these don't seem to provide additional information 
- format for readability and consistency

```
            Map<VariableReferenceExpression, RowExpression> dereferencesMap = expressions.inverse().entrySet().stream()
                    .collect(toImmutableMap(Map.Entry::getKey, entry -> castToRowExpression(entry.getValue())));
            Assignments assignments = Assignments.builder()
                    .putAll(identityAssignmentsAsSymbolReferences(unnestNode.getSource().getOutputVariables()))
                    .putAll(dereferencesMap)
                    .build();
            ProjectNode source = new ProjectNode(context.getIdAllocator().getNextId(), unnestNode.getSource(), assignments);

            return Result.ofPlanNode(
                    new UnnestNode(context.getIdAllocator().getNextId(),
                            source,
                            ImmutableList.<VariableReferenceExpression>builder()
                                    .addAll(unnestNode.getReplicateVariables())
                                    .addAll(expressions.values())
                                    .build(),
                            unnestNode.getUnnestVariables(),
                            unnestNode.getOrdinalityVariable()));
```

File: presto-main/src/test/java/com/facebook/presto/sql/planner/TestDereferencePushDown.java [hunk 19]
```diff
@@ -0,0 +1,203 @@
...
+import static com.facebook.presto.sql.planner.assertions.PlanMatchPattern.unnest;
+import static com.facebook.presto.sql.planner.assertions.PlanMatchPattern.values;
+import static com.facebook.presto.sql.planner.plan.ExchangeNode.Scope.LOCAL;
+import static com.facebook.presto.sql.planner.plan.ExchangeNode.Type.GATHER;
+import static com.facebook.presto.sql.planner.plan.ExchangeNode.Type.REPARTITION;
+import static com.facebook.presto.sql.planner.plan.JoinNode.Type.INNER;
+import static com.facebook.presto.sql.tree.SortItem.NullOrdering.LAST;
+import static com.facebook.presto.sql.tree.SortItem.Ordering.ASCENDING;
+
+public class TestDereferencePushDown
```
Comment (mbasmanova) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/TestDereferencePushDown.java:44]: I ran this test with coverage and I see some code paths in PushDownDereferences rule not covered. Would you add tests to ensure full coverage of the new rule?

File: presto-main/src/test/java/com/facebook/presto/sql/planner/TestDereferencePushDown.java [hunk 20]
```diff
@@ -0,0 +1,203 @@
...
+        extends BasePlanTest
+{
+    public TestDereferencePushDown()
+    {
+        super(ImmutableMap.of(SystemSessionProperties.PUSHDOWN_SUBFIELDS_ENABLED, "true"));
+    }
+
+    @Test
+    public void testDereferencePushdownJoin()
```
Comment (mbasmanova) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/TestDereferencePushDown.java:53]: nit: consider dropping `DereferencePushdown` from test names for brevity, e.g. testJoin, testUnnest, etc.

File: presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/RowExpressionVerifier.java [hunk 21]
```diff
@@ -322,6 +326,28 @@ protected Boolean visitBooleanLiteral(BooleanLiteral expected, RowExpression act
...
+        SpecialFormExpression actualDereference = (SpecialFormExpression) actual;
+        if (actualDereference.getArguments().size() == 2 &&
+                actualDereference.getArguments().get(0).getType() instanceof RowType &&
+                actualDereference.getArguments().get(1) instanceof ConstantExpression) {
+            RowType rowType = (RowType) actualDereference.getArguments().get(0).getType();
+            Object value = LiteralInterpreter.evaluate(TEST_SESSION.toConnectorSession(), (ConstantExpression) actualDereference.getArguments().get(1));
+            Preconditions.checkState(value instanceof Long);
+            long index = (Long) value;
+            Preconditions.checkState(index >= 0 && index < rowType.getFields().size());
+            RowType.Field field = rowType.getFields().get((int) index);
```
Comment (mbasmanova) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/RowExpressionVerifier.java:344]: use Math.toIntExact

File: presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/RowExpressionVerifier.java [hunk 22]
```diff
@@ -322,6 +326,28 @@ protected Boolean visitBooleanLiteral(BooleanLiteral expected, RowExpression act
...
+        if (!(actual instanceof SpecialFormExpression) || !(((SpecialFormExpression) actual).getForm().equals(DEREFERENCE))) {
+            return false;
+        }
+        SpecialFormExpression actualDereference = (SpecialFormExpression) actual;
+        if (actualDereference.getArguments().size() == 2 &&
+                actualDereference.getArguments().get(0).getType() instanceof RowType &&
+                actualDereference.getArguments().get(1) instanceof ConstantExpression) {
+            RowType rowType = (RowType) actualDereference.getArguments().get(0).getType();
+            Object value = LiteralInterpreter.evaluate(TEST_SESSION.toConnectorSession(), (ConstantExpression) actualDereference.getArguments().get(1));
+            Preconditions.checkState(value instanceof Long);
```
Comment (mbasmanova) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/RowExpressionVerifier.java:341]: static import Preconditions.checkState

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushDownDereferences.java (hunk 5 above)
Comment (zhenxiao) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushDownDereferences.java]: fixed

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushDownDereferences.java (hunk 6 above)
Comment (zhenxiao) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushDownDereferences.java]: fixed

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushDownDereferences.java (hunk 8 above)
Comment (zhenxiao) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushDownDereferences.java]: fixed

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushDownDereferences.java (hunk 9 above)
Comment (zhenxiao) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushDownDereferences.java]: fixed

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushDownDereferences.java (hunk 10 above)
Comment (zhenxiao) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushDownDereferences.java]: fixed

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushDownDereferences.java (hunk 11 above)
Comment (zhenxiao) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushDownDereferences.java]: fixed

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushDownDereferences.java (hunk 12 above)
Comment (zhenxiao) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushDownDereferences.java]: fixed

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushDownDereferences.java (hunk 13 above)
Comment (zhenxiao) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushDownDereferences.java]: fixed

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushDownDereferences.java (hunk 14 above)
Comment (zhenxiao) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushDownDereferences.java]: fixed

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushDownDereferences.java (hunk 15 above)
Comment (zhenxiao) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushDownDereferences.java]: fixed

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushDownDereferences.java (hunk 16 above)
Comment (zhenxiao) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushDownDereferences.java]: fixed

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushDownDereferences.java (hunk 17 above)
Comment (zhenxiao) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushDownDereferences.java]: fixed

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushDownDereferences.java (hunk 18 above)
Comment (zhenxiao) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushDownDereferences.java]: fixed

File: presto-main/src/test/java/com/facebook/presto/sql/planner/TestDereferencePushDown.java (hunk 20 above)
Comment (zhenxiao) [reply presto-main/src/test/java/com/facebook/presto/sql/planner/TestDereferencePushDown.java]: fixed

File: presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/RowExpressionVerifier.java (hunk 21 above)
Comment (zhenxiao) [reply presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/RowExpressionVerifier.java]: fixed

File: presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/RowExpressionVerifier.java (hunk 22 above)
Comment (zhenxiao) [reply presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/RowExpressionVerifier.java]: fixed

**Comment by zhenxiao** [comment]
to be continued in: https://github.com/prestodb/presto/pull/14829

**Pull request closed without merging**